docs/schulcloud-server/architecture.md

```diff
@@ -0,0 +1,146 @@
+# Software Architecture
+
+## Goals
+
+Our architecture aims to achieve the following goals:
+
+- Maintainability
+  - it should be easy as possible to make changes that do not change the behaviour of the system (refactoring)
+  - it should be easy to exchange entire components of the system, without impact on other components.
+- Extendability
+  - it should be easy to add new functionality to the system
+- Agility
+  - it should be easy to react to changing requirements during our development process
+- Change Security
+- it should be easy to determine the correctness of the system after making any changes
+
+## Principles
+
+In order to achieve these goals, we try to follow the principles detailed below.
+These principles apply to all layers of our software, from lines of code and methods to modules and architectural layers.
+
+- **Single Responsibility / Seperation of Concerns**
+  - each piece of code should have a single layer of abstraction/detail
+  - each piece of code should have a single reason to change
+- **Open/Closed Principle**
+  - design to be open to extension, but closed to modification
+  - **Liskov Substitution**
+  - the specific input may be more generic than its interface
+  - the specific output may be more specialized than its interface
+- **Interface Segregation**
+  - multiple small interfaces are preferred over big interfaces
+- **Dependency Inversion Principle**
+  - always depend on interfaces, not implementations
+  - higher level parts should not depend on lower level parts.
+- **Keep It Simple (KISS)**
+  - any piece of code should be simple and readable
+  - any logic should be broken down to be trivial
+  - beware of overenginiering and premature optimisation
+- **You Aint Gonna Need It (YAGNI)**
+  - keep decisions open for as long as possible
+  - build only what you need to build, stay flexible for future requirements
+- **Do Not Repeat Yourself (DRY)**
+  - do not solve the same responsability or concern in multiple places
+  - beware of things that look similar, but are not. for example, things that change for different reasons should not be combined, even if their code looks the same
+
+## Server Layer Architecture
+
+We generally distinguish three different layers in our server architecture: The API Layer, the Repository Layer, and the Domain Layer.
+
+![Architecture Layers](./img/clean-architecture-layers.png)
+
+Note that based on the Dependency Inversion Principle, the Domain Layer does not have any dependencies. Instead, both the API and Repository Layer depend on its abstractions.
+
+### Domain Layer
+
+The Domain Layer contains the business logic of the application. As mentioned above, it is not allowed to know about anything outside the domain layer itself.
```

Comment:
Splitting UC and service in two layer. Nothing from UC should be importat into service layer.